internal/codespaces/api/api.go

```diff
@@ -301,24 +303,24 @@ func findNextPage(linkValue string) string {
 // If the codespace is not found, an error is returned.
 // If includeConnection is true, it will return the connection information for the codespace.
 func (a *API) GetCodespace(ctx context.Context, codespaceName string, includeConnection bool) (*Codespace, error) {
-	req, err := http.NewRequest(
-		http.MethodGet,
-		a.githubAPI+"/user/codespaces/"+codespaceName,
-		nil,
-	)
-	if err != nil {
-		return nil, fmt.Errorf("error creating request: %w", err)
-	}
-
-	if includeConnection {
-		q := req.URL.Query()
-		q.Add("internal", "true")
-		q.Add("refresh", "true")
-		req.URL.RawQuery = q.Encode()
-	}
-
-	a.setHeaders(req)
-	resp, err := a.do(ctx, req, "/user/codespaces/*")
+	resp, err := a.withRetry(func() (*http.Response, error) {
```

Comment:
Couldn't we avoid making any changes to methods like `GetCodespace` or `StartCodespace` and implement retries transparently within `a.do()`? That way the retry mechanism could just kick in for any request that we consider idempotent (typically, GET requests or POST to the `start` endpoint).

Reply:
@mislav the problem with `a.do()` is two fold: 

1. Makes it harder to control _which_ methods we want to retry.
2. More importantly, if the server ever returns redirects, `a.do()` does not recreate the request therefore net/http will not be able to send the request body again in the case of POST requests. FYI: the net/http library _does_ resend the body in the case of `*bytes.Buffer` and `*bytes/strings.Reader`, but that cannot be guaranteed that a caller won't use another underlying `io.Reader` where the server sends a redirect response.

Reply:
@mislav friendly ping 👋🏼

Reply:
Great point about the difficulties of retrying requests that have a body.